### PR TITLE
Fix how we compute the final non-padding token for ForSequenceClassification models

### LINKED ISSUE #30004: Gemma Tokenizer add padding to the left, but the code to find sequence length assume pad tokens are added to the right
Gemma Tokenizer add padding to the left, but the code to find sequence length assume pad tokens are added to the right

Reference:
https://github.com/huggingface/transformers/blob/97c00cdfe132164dbd793447a088432fa359fd36/src/transformers/models/gemma/modeling_gemma.py#L1318-L1320

### LINKED ISSUE #35352: Maybe the way SequenceClassification Model calculates the last non-pad token is not reasonable.
### System Info

None

### Who can help?

@ArthurZucker 

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

SequenceClassification Model  finds the last token that is not a padding token in each row  by cal the first position of pad token when a `pad_token_id` is defined in the configuration. 
the code in https://github.com/huggingface/transformers/blob/main/src/transformers/models/llama/modeling_llama.py#L951

```
sequence_lengths = torch.eq(input_ids, self.config.pad_token_id).int().argmax(-1) - 1
```

When pad_token is eos token, what is calculated using this method will be the position of the first eos token. However, in some LLM templates, eos is often added at the end of the prompt. Also, when a model such llama does not have a default pad token, using eos as a pad token is a common practice.

For example, in llama, eos token is <|eot_id|>

```
from transformers import AutoModelForSequenceClassification, AutoTokenizer
model = AutoModelForSequenceClassification.from_pretrained("meta-llama/Llama-3.2-1B-Instruct", num_labels=1)
tokenizer = AutoTokenizer.from_pretrained("meta-llama/Llama-3.2-1B-Instruct")
tokenizer.eos_token, model.config.pad_token_id, tokenizer.pad_token
# ('<|eot_id|>', None, None)
```
Using chat template to encode QA pairs we can get

```
prompt = "what is 1+1?"
response1 = "1+1=2"
response2 = "1+1=3"

conv1 = [{"role": "user", "content": prompt}, {"role": "assistant", "content": response1}]
conv2 = [{"role": "user", "content": prompt}, {"role": "assistant", "content": response2}]

conv1_tokenized = tokenizer.apply_chat_template(conv1, tokenize=True, return_tensors="pt")
conv2_tokenized = tokenizer.apply_chat_template(conv2, tokenize=True, return_tensors="pt")

# conv1_tokenized(input_ids)
# (tensor([[128000, 128006,    882, 128007,    271,  12840,    374,    220,     16,
#             10,     16,     30, 128009, 128006,  78191, 128007,    271,     16,
#              10,     16,     28,     17, 128009, 128006,  78191, 128007,    271]]),
# conv1_tokenized(tokens)
# '<|begin_of_text|><|start_header_id|>user<|end_header_id|>\n\nwhat is 1+1?<|eot_id|><|start_header_id|>assistant<|end_header_id|>\n\n1+1=2<|eot_id|><|start_header_id|>assistant<|end_header_id|>\n\n')
```
In Llama chat template, eos_token is added on the end of prompt. If we do not pad the input_ids and don't set pad_token_id, the score is correct. 

```
with torch.no_grad():
    score1 = model(conv1_tokenized).logits[0][0].item()
    score2 = model(conv2_tokenized).logits[0][0].item()
print(f"Score for response 1: {score1}")
print(f"Score for response 2: {score2}")
# Score for response 1: 1.7297050952911377
# Score for response 2: 1.43972647190094
```

If we set pad_token_id = eos_token_id, we get the same score for different QA pairs with the same prompt

```
model.config.pad_token_id = tokenizer.eos_token_id
with torch.no_grad():
    score1 = model(conv1_tokenized).logits[0][0].item()
    score2 = model(conv2_tokenized).logits[0][0].item()
print(f"Score for response 1: {score1}")
print(f"Score for response 2: {score2}")
# Score for response 1: -1.857212781906128
# Score for response 2: -1.857212781906128
```

This is because the score of SequenceClassification Model is the last non-pad token's logits，and the last non-pad token is the token before the first eos token. This is incorrect especially when training reward models with preference pairs that have the same prompt.

The complete recurrence script is as follows

```
from transformers import AutoModelForSequenceClassification, AutoTokenizer
model = AutoModelForSequenceClassification.from_pretrained("meta-llama/Llama-3.2-1B-Instruct", num_labels=1)
tokenizer = AutoTokenizer.from_pretrained("meta-llama/Llama-3.2-1B-Instruct")
prompt = "what is 1+1?"
response1 = "1+1=2"
response2 = "1+1=3"

conv1 = [{"role": "user", "content": prompt}, {"role": "assistant", "content": response1}]
conv2 = [{"role": "user", "content": prompt}, {"role": "assistant", "content": response2}]

conv1_tokenized1 = tokenizer.apply_chat_template(conv1, tokenize=True, return_tensors="pt")
conv2_tokenized1 = tokenizer.apply_chat_template(conv2, tokenize=True, return_tensors="pt")
model.config.pad_token_id = tokenizer.eos_token_id
with torch.no_grad():
    score1 = model(conv1_tokenized1).logits[0][0].item()
    score2 = model(conv2_tokenized1).logits[0][0].item()
print(f"Score for response 1: {score1}")
print(f"Score for response 2: {score2}")
```

### Expected behavior

Maybe we should prioritize using attention mask to calculate  the  position of the last non-pad token.

### LINKED ISSUE #35909: ForSequenceClassification models assume right-padding
We have `ForSequenceClassification` variants for several of our CLMs, and this line is commonly copied between them:

`sequence_lengths = torch.eq(input_ids, self.config.pad_token_id).int().argmax(-1) - 1`

However, this line assumes **right-padding**, because it treats the length of the sequence as the number of tokens before the first pad token (when multiple values have the max value, then `argmax` returns the first index with that value).

If we have left-padding instead, then this will break because it will compute a sequence length of -1. This was reported as a bug in Gemma [here](https://github.com/huggingface/transformers/issues/30004), but it likely affects other models as well.

`tokenizer.padding_side` is probably not accessible to the model, because it is stored in `tokenizer_config` and not `config`. As a result, I think the solution here will have to be rewriting this code so that it can handle either left- or right-padding.

### PR DESCRIPTION
**Pull request by Rocketknight1**
We have a lot of CLM models with `ForSequenceClassification` heads. These models are supposed to use the hidden state at the final non-padding token as the input to the classification head. However, the way they compute it is a bit weird - they get the index of the leftmost token that is equal to `pad_token_id` and subtract 1 from it. This has a few issues:

- It breaks on left-padding
- It creates index arithmetic issues when `pad_token_id` is absent, that need workarounds
- It depends on implementation details of `argmax()`, specifically that when multiple indices have the same maximum value, it always returns the smallest one

This PR replaces that logic with simpler logic that actually searches for what we want, the **rightmost non-padding token**, not the **token next to the leftmost padding token**. This means the same logic works with left-padding, right-padding, no-padding, or even padding on both sides (I don't think any models do that, but we're ready if they do!)

Fixes #30004
Fixes #35352
Fixes #35909

### CODE REVIEW

**Comment by Rocketknight1** [comment]
cc @ArthurZucker for core maintainer review - but if you have too much to do, let me know and I'll find another reviewer!

**Comment by Rocketknight1** [comment]
cc @Cyrilvallez actually, as core-maintainer-in-training

File: src/transformers/models/gemma/modeling_gemma.py [hunk 1]
```diff
@@ -948,17 +948,21 @@ def forward(
...
+                non_pad_mask = torch.ne(input_ids, self.config.pad_token_id).int().to(logits.device)
+                token_indices = torch.arange(input_ids.shape[-1], device=logits.device)
+                last_non_pad_token = (token_indices * non_pad_mask).max(-1).values
             else:
-                sequence_lengths = -1
+                last_non_pad_token = -1
+                logger.warning_once(
+                    f"{self.__class__.__name__} will not detect padding tokens in `inputs_embeds`. Results may be "
+                    "unexpected if using padding tokens in conjunction with `inputs_embeds.`"
+                )
```
Comment (Cyrilvallez) [inline src/transformers/models/gemma/modeling_gemma.py:963]: Instead of adding the warning everywhere, maybe we should do it the other way around and actually remove it from the few classes where it is present? It is a fairly niche case, and people using it should be aware of the caveat IMO. It's nice to have less warnings in general, WDYT? (It would allow to simplify branching as well)

**Review by Cyrilvallez — COMMENTED** [review]
Hey! Nicely done, indeed much simpler than before! Just added some small comments! Let me know what you think!

File: src/transformers/models/bloom/modeling_bloom.py [hunk 2]
```diff
@@ -1122,21 +1122,21 @@ def forward(
...
-            sequence_lengths = -1
+            last_non_pad_token = -1
         else:
             if input_ids is not None:
-                # if no pad token found, use modulo instead of reverse indexing for ONNX compatibility
-                sequence_lengths = torch.eq(input_ids, self.config.pad_token_id).int().argmax(-1) - 1
-                sequence_lengths = sequence_lengths % input_ids.shape[-1]
-                sequence_lengths = sequence_lengths.to(logits.device)
+                # To handle both left- and right- padding, we take the rightmost token that is not equal to pad_token_id
+                non_pad_mask = torch.ne(input_ids, self.config.pad_token_id).int().to(logits.device)
```
[the remark this answers is not in the document]
Comment (Rocketknight1) [reply src/transformers/models/bloom/modeling_bloom.py]: Did this, and also merged `.int()` and `.to()` into a single `.to()`!

File: src/transformers/models/bloom/modeling_bloom.py [hunk 3]
```diff
@@ -1122,21 +1122,21 @@ def forward(
...
+                token_indices = torch.arange(input_ids.shape[-1], device=logits.device)
+                last_non_pad_token = (token_indices * non_pad_mask).max(-1).values
```
[the remark this answers is not in the document]
Comment (Rocketknight1) [reply src/transformers/models/bloom/modeling_bloom.py]: This isn't an `argmax`! It's actually a fake `argmax` where I make a masked index array and then compute the `max`. It just looks a lot like an argmax, lol

File: src/transformers/models/gemma/modeling_gemma.py (hunk 1 above)
Comment (Rocketknight1) [reply src/transformers/models/gemma/modeling_gemma.py]: Hmm - I think a lot of classes actually support `inputs_embeds` as well as `input_ids`, so we still need this on most classes I think! It just only fires when users actually pass `inputs_embeds` and not `input_ids`.

File: src/transformers/models/bloom/modeling_bloom.py [hunk 4]
```diff
@@ -1122,21 +1122,21 @@ def forward(
         if self.config.pad_token_id is None and batch_size != 1:
             raise ValueError("Cannot handle batch sizes > 1 if no padding token is defined.")
         if self.config.pad_token_id is None:
-            sequence_lengths = -1
+            last_non_pad_token = -1
```
[the remark this answers is not in the document]
Comment (Rocketknight1) [reply src/transformers/models/bloom/modeling_bloom.py:1125]: I simplified the if/else here a lot, yeah!

File: src/transformers/models/ctrl/modeling_tf_ctrl.py [hunk 5]
```diff
@@ -868,42 +868,33 @@ def call(
...
+            if self.config.pad_token_id is None and logits_shape[0] != 1:
                 raise ValueError("Cannot handle batch sizes > 1 if no padding token is defined.")
 
-            if not tf.is_tensor(sequence_lengths):
-                in_logits = logits[0:batch_size, sequence_lengths]
-
-            loss = self.hf_compute_loss(tf.reshape(labels, [-1, 1]), tf.reshape(in_logits, [-1, self.num_labels]))
-
-        pooled_logits = in_logits if in_logits is not None else logits
+            loss = self.hf_compute_loss(tf.reshape(labels, [-1]), tf.reshape(pooled_logits, [-1, self.num_labels]))
```
[the remark this answers is not in the document]
Comment (Rocketknight1) [reply src/transformers/models/ctrl/modeling_tf_ctrl.py:897]: Yes, this should be the same or more correct! The loss is tested by the `test_pt_tf_equivalence` tests, so we should see if they go out of alignment

File: src/transformers/models/gemma/modeling_gemma.py (hunk 1 above)
Comment (Cyrilvallez) [reply src/transformers/models/gemma/modeling_gemma.py]: Be aware that it breaks `torch.compile` as well, would anyone want to use it! But once again, it is quite a niche case, and it's already here for some models, so I'll let you judge -- we can keep it if you think that removing it would bring confusion/error for users 🤗

File: src/transformers/models/bloom/modeling_bloom.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Rocketknight1) [reply src/transformers/models/bloom/modeling_bloom.py]: True, yes! I could just use argmax instead

File: src/transformers/models/gemma/modeling_gemma.py (hunk 1 above)
Comment (Rocketknight1) [reply src/transformers/models/gemma/modeling_gemma.py]: I think it's better to leave the warning, because the results will generally be wrong in that case. In future, we might consider raising an error and asking users to supply an `attention_mask` in that case!

**Review by Cyrilvallez — APPROVED** [review]
All right, LGTM! Thank you very much!
